src/validation.cpp

```diff
@@ -399,6 +399,35 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
     LimitMempoolSize(mempool, GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60);
 }
 
+// Used to avoid mempool polluting consensus critical paths if CCoinsViewMempool
+// were somehow broken and returning the wrong scriptPubKeys
+static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &view, CTxMemPool& pool,
+                 unsigned int flags, bool cacheSigStore, PrecomputedTransactionData& txdata) {
+    AssertLockHeld(cs_main);
+
+    assert(!tx.IsCoinBase());
+    {
+        LOCK(pool.cs);
+        for (const CTxIn& txin : tx.vin) {
+            const Coin& coin = view.AccessCoin(txin.prevout);
+            if (coin.IsSpent()) return false;
```

Comment:
In commit "Add CheckInputs wrapper CCoinsViewMemPool -> non-consensus-critical"

Could you add a comment explaining the return? It's not clear to me if this is supposed to be an optimization or a paranoid safety check. If it's meant to be the latter, maybe a way to make it more paranoid would be to do `if (coin.IsSpent()) has_spent_input = true` here and assert `!CheckInputs || !has_spent_input` at the end.